terraform.tf

```diff
@@ -0,0 +1,17 @@
+terraform {
```

Comment:
I believe it's @fraenkel's position that this is solely the responsibility of the consumer of this module

Reply:
1. The issue is that the resources declaration is somewhat tied to a particular version of the provider,  there were fields in the resources declaration that are not backwards compatible with newer version, 
example.
the `location/zone` parameter  changed from one version of the provider to the next.

```
resource "google_filestore_instance" "instance" {
  zone = "us-central1-b"
...
```
after 3.9.0
```
resource "google_filestore_instance" "instance" {
  location = "us-central1-b"
...
}
```

2. It was caught by `tflint`
```
Warning: Missing version constraint for provider "google-beta" in "required_providers" (terraform_required_providers)

on [gcr.tf](http://gcr.tf/) line 1:

1: provider "google-beta" {

Reference: https://github.com/terraform-linters/tflint/blob/v0.35.0/docs/rules/terraform_required_providers.md

Warning: Missing version constraint for provider "google" in "required_providers" (terraform_required_providers)

on [main.tf](http://main.tf/) line 17:

17: provider "google" {

Reference: https://github.com/terraform-linters/tflint/blob/v0.35.0/docs/rules/terraform_required_providers.md

Warning: Missing version constraint for provider "random" in "required_providers" (terraform_required_providers)

on [main.tf](http://main.tf/) line 26:

26: resource "random_uuid" "id" {}

Reference: https://github.com/terraform-linters/tflint/blob/v0.35.0/docs/rules/terraform_required_providers.md
```

3. google official gke terraform module for [reference](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/versions.tf)

In conclusion the absence of the version pinning tasks the customer with the burden of `figuring` out on which version the module "works".

Reply:
I would agree if all the modules are declared and maintained on the same repo but not on remotely referenced  modules, which is the goal and the standard(can be seen on aws/google modules `versions.tf`)

from the footprint of the implicit providers block
```
Note: Only provider configurations are inherited by child modules, not provider source or version requirements. Each module must [declare its own provider requirements](https://www.terraform.io/language/providers/requirements). This is especially important for non-HashiCorp providers.
```

Reply:
@fraenkel note the goal is to comply with industry standards and publish this module on the terraform registry.

Reply:
So even tflint has a comment here:
https://github.com/terraform-linters/tflint/blob/master/docs/rules/terraform_required_providers.md#how-to-fix
```
This approach is now discouraged, particularly for child modules.
```

Reply:
child modules are not used here.

Reply:
it is given its a child module to platform-apps.

Reply:
What I meant was that this module does not have any child modules. Also this module is not considered a child module to platform apps unless platform apps is another module.

[child-modules](https://www.terraform.io/language/modules#child-modules)
```
A module that has been called by another module is often referred to as a child module.
```

Reply:
Everything is a module, the top level is called the root module.
https://www.terraform.io/language/modules#the-root-module

The point is that child modules are invoked via a module block which is how this is invoked.

Reply:
The issue with provider versions is that if we aren't careful we are going to have 2 sets, one from platform-apps and one from the module which creates a total mess. Again our use case is that everything is driven from platform-apps which means it should be defining the provider versions and the providers used by modules.

Reply:
if we want a module that only works with platform-apps then we we should stick it in a folder on its repo. But this is public published terraform module and as is any client that tries to consume it needs to figure out the right versions in order to use it(not only the providers but the terraform version), again see the use-case I noted.

Reply:
we can have different version constraints as long as it produces a union. As for 4.x, It was not part of the scope of the story, but we can add it.

Reply:
see how easy it is to add otherwise its just more work back and forth.